File: AutoPlot/App/Utils/Storage/Core/s3_storage_backend.py
import tempfile
import uuid
from pathlib import Path

import boto3
from botocore.client import Config

from App.Utils.Storage.Core.storage_exceptions import S3OperationException


class S3StorageBackend:
    TEMP_DIR = Path(tempfile.gettempdir())

    @staticmethod
    def temp_path_for_key(key: str):
        ext = Path(key).suffix
        file_name = f"s3reader-{uuid.uuid4()}{ext}"
        return S3StorageBackend.TEMP_DIR / file_name

    @staticmethod
    def create_client(endpoint: str | None):
        if endpoint:
            return boto3.client(
                "s3",
                endpoint_url=endpoint,
                aws_access_key_id="test",
                aws_secret_access_key="test",
                region_name="us-east-1",
                config=Config(s3={'addressing_style': 'path'})
            )
        return boto3.client("s3")

    @staticmethod
    def exists(bucket: str, key: str, endpoint: str | None = None) -> bool:
        client = S3StorageBackend.create_client(endpoint)

        try:
            client.head_object(Bucket=bucket, Key=key)
            return True
        except Exception:
            pass

        if not key.endswith("/"):
            prefix = key + "/"
        else:
            prefix = key

        try:
            resp = client.list_objects_v2(
                Bucket=bucket,
                Prefix=prefix,
                MaxKeys=1
            )
            return "Contents" in resp
        except Exception as ex:
            print(ex)
            return False

    @staticmethod
    def read(bucket: str, key: str, endpoint: str | None = None):
        client = S3StorageBackend.create_client(endpoint)
        tmp = S3StorageBackend.temp_path_for_key(key)

        try:
            client.download_file(bucket, key, str(tmp))
            return tmp
        except Exception as ex:
            raise S3OperationException(bucket, key, ex)

    @staticmethod
    def read_directory(bucket: str, prefix: str, endpoint: str | None = None):
        client = S3StorageBackend.create_client(endpoint)

        target_dir = tempfile.mkdtemp()

        paginator = client.get_paginator("list_objects_v2")
        pages = paginator.paginate(Bucket=bucket, Prefix=prefix)

        for page in pages:
            if "Contents" not in page:
                continue

            for obj in page["Contents"]:
                key = obj["Key"]
                if key.endswith("/"):
                    continue

                dst = Path(target_dir) / Path(key).name
                try:
                    client.download_file(bucket, key, str(dst))
                except Exception as ex:
                    raise S3OperationException(bucket, key, ex)

        return target_dir

    @staticmethod
    def write(bucket: str, key: str, local_path: Path, endpoint: str | None = None):
        client = S3StorageBackend.create_client(endpoint)

        try:
            client.upload_file(str(local_path), bucket, key)
        except Exception as ex:
            raise S3OperationException(bucket, key, ex)

File: big-data-core/modules/utils/src/main/scala/Utils/Storage/Core/S3StorageBackend.scala
package Utils.Storage.Core

import software.amazon.awssdk.auth.credentials.{AwsBasicCredentials, StaticCredentialsProvider}
import software.amazon.awssdk.regions.Region
import software.amazon.awssdk.services.s3.S3Client
import software.amazon.awssdk.services.s3.model._

import java.net.URI
import java.nio.file.{Files, Path, Paths, StandardCopyOption}
import java.util.UUID
import scala.jdk.CollectionConverters.CollectionHasAsScala

/**
 * S3 storage backend utilities. Supports reading single objects, recursive
 * directory-like downloads and writing objects. When `endpoint` is provided
 * a dummy client with static credentials is created (useful for localstack).
 */
object S3StorageBackend {

  private val tempDirBase: Path = Paths.get(System.getProperty("java.io.tmpdir"))

  /**
   * Create a lightweight S3 client pointing to a custom `endpoint` (used for testing/localstack).
   *
   * @param endpoint endpoint URL string
   * @return configured `S3Client`
   */
  private def createDummyClient(endpoint: String): S3Client = {
    S3Client.builder()
      .region(Region.US_EAST_1)
      .endpointOverride(URI.create(endpoint))
      .credentialsProvider(
        StaticCredentialsProvider.create(AwsBasicCredentials.create("test", "test"))
      )
      .forcePathStyle(true)
      .build()
  }

  /**
   * Create a temporary path for a given S3 `key`, preserving file extension.
   *
   * @param key S3 object key
   * @return temporary `Path` on the local filesystem
   */
  private def tempPathForKey(key: String): Path = {
    val extension = {
      val name = new java.io.File(key).getName
      val dotIndex = name.lastIndexOf(".")
      if (dotIndex != -1) name.substring(dotIndex) else ""
    }

    val fileName = s"s3reader-${UUID.randomUUID()}$extension"

    tempDirBase.resolve(fileName)
  }

  /**
   * Read an object from S3 into a temporary `Path` and return it.
   *
   * @param bucket S3 bucket name
   * @param key S3 object key
   * @param endpoint optional custom endpoint (localstack/test)
   * @return temporary `Path` where the downloaded object is stored
   * @throws S3OperationException on download errors
   */
  def read(bucket: String, key: String, endpoint: Option[String] = None): Path = {
    val client: S3Client = endpoint match {
      case Some(uri) => createDummyClient(uri)
      case None => S3Client.create()
    }

    val tmp = tempPathForKey(key)

    val req = GetObjectRequest.builder()
      .bucket(bucket)
      .key(key)
      .build()

    try {
      client.getObject(req, tmp)
    } catch {
      case exception: Exception => throw new S3OperationException(bucket, key, exception)
    }
    tmp
  }

  /**
   * Recursively download a subset of an S3 "directory" (prefix) to a temporary directory.
   * The `includeDirs` elements must start with the root folder name and are used to
   * filter which keys are downloaded.
   *
   * @param bucket S3 bucket name
   * @param key prefix representing the root key to download
   * @param endpoint optional custom endpoint
   * @param includeDirs sequence of include paths (must start with the root name)
   * @return temporary `Path` containing the downloaded subset
   */
  def readDirectoryRecursive(
    bucket: String,
    key: String,
    endpoint: Option[String] = None,
    includeDirs: Seq[String]
  ): Path = {

    val client = endpoint match {
      case Some(uri) => S3StorageBackend.createDummyClient(uri)
      case None     => S3Client.create()
    }

    val normalizedKey = key.stripSuffix("/")
    val rootName = Paths.get(normalizedKey).getFileName.toString
    val tempDirBasePath = Files.createTempDirectory(tempDirBase, null)
    val tempDir = tempDirBasePath.resolve(rootName)
    Files.createDirectories(tempDir)

    val includePaths: Set[String] = includeDirs.map { p =>
      val normalized = p.replace('\\', '/').stripPrefix("/").stripSuffix("/")
      val parts = normalized.split("/").toList

      parts match {
        case head :: tail if head == rootName =>
          tail.mkString("/")
        case _ =>
          throw new IllegalArgumentException(
            s"includeDir must start with root directory '$rootName': $p"
          )
      }
    }.toSet

    /**
     * Recursive helper that lists objects using `ListObjectsV2` and downloads
     * each matching object into the temporary directory.
     *
     * @param continuationToken optional continuation token for paged listing
     */
    def listAndDownload(continuationToken: Option[String]): Unit = {
      val reqBuilder = ListObjectsV2Request.builder()
        .bucket(bucket)
        .prefix(normalizedKey)

      continuationToken.foreach(reqBuilder.continuationToken)
      val resp = client.listObjectsV2(reqBuilder.build())

      resp.contents().asScala.foreach { obj =>
        val relative = obj.key()
          .stripPrefix(normalizedKey)
          .stripPrefix("/")
          .replace("/", java.io.File.separator)

        if (relative.nonEmpty) {
          val includeThis =
            includePaths.isEmpty ||
              includePaths.exists(dir => relative.replace("\\", "/").startsWith(dir))

          if (includeThis) {
            val targetPath = tempDir.resolve(relative)
            Option(targetPath.getParent).foreach(Files.createDirectories(_))

            val tmpFile = S3StorageBackend.read(bucket, obj.key(), endpoint)
            Files.copy(tmpFile, targetPath, StandardCopyOption.REPLACE_EXISTING)
          }
        }
      }

      Option(resp.nextContinuationToken()).foreach(token => listAndDownload(Some(token)))
    }

    listAndDownload(None)
    tempDir
  }

  /**
   * Write a local file to S3 under the provided `key`.
   *
   * @param bucket S3 bucket name
   * @param key S3 object key
   * @param localPath local file `Path` to upload
   * @param endpoint optional custom endpoint
   */
  def write(bucket: String, key: String, localPath: Path, endpoint: Option[String] = None): Unit = {
    val client: S3Client = endpoint match {
      case Some(uri) => createDummyClient(uri)
      case None => S3Client.create()
    }
    val req = PutObjectRequest.builder()
      .bucket(bucket)
      .key(key)
      .build()

    try {
      client.putObject(req, localPath)
    } catch {
      case exception: Exception => throw new S3OperationException(bucket, key, exception)
    }
  }
}